Simplify and fix the shadow store reservation on Win64

dra27 opened the pull request:
This PR addresses two Win64 ABI violations in the native backend/compiler (Win64 in this case means Cygwin64 - MSVC64 - mingw-w64). The first was indirectly spotted while fixing the second, which is that OCaml 5's calling convention clobbers `rbx` which is recorded in the Unix register descriptions but accidentally wasn't in the Windows ones. This predominantly affects large-argument and noalloc function calls, so I think we've just been "lucky" in not seeing that so far.

The second part is more complicated, and relates to the [x64 calling convention on Windows](https://learn.microsoft.com/en-us/cpp/build/x64-calling-convention?view=msvc-170). This requires that the _caller_ reserve space for the four argument registers to be spilled by the function. The callee is not _required_ to use this space for spilling the registers, but the space is supposed to be available, regardless of whether the callee even has 4 parameters.

ocamlopt generates calls to C functions in three different ways:
1. Via `caml_c_call`. This is used when the function being called has 4 or fewer arguments. It's implemented in `amd64.S` and is called with the arguments registers already initialised and the function address in `rax`.
2. Via `caml_c_call_stack_args`. This is used when the function being called has more than 4 arguments. As for `caml_c_call`, the argument registers are already initialised, but the additional arguments are on the wrong stack, as they are on the OCaml stack. `caml_c_call_stack_args` therefore copies these to the C stack and then transfers control to the function address in `rax`.
3. Directly. This is used only for noalloc calls with 4 arguments or fewer. `rbx` is clobbered to maintain the OCaml stack pointer across the call.

OCaml 5.0 reserved 32 bytes on the C stack just prior to the C call for cases 1 and 2 but the shadow store was not reserved at all for the direct call (case 3). Mostly, this had slipped through undetected - although I'm 99% certain this is the root cause of a bug seen by @polytypic with the eio library on mingw-w64. This mistake was trivially spotted in #11835, as compiling with `-Od` (cl's equivalent of gcc's `-O0`) generates function prologues which always spill argument registers to their home locations, causing very easy-to-detect stack corruption for msvc64 programs.

The simple fix initially put in #11835 was to reserve the required 32 bytes of stack prior to the `call` for the noalloc case, but this isn't entirely satisfactory. OCaml 4.x is (of course) aware of the shadow store requirement, and this stack _was_ being reserved **in all three cases**. The problem is that the reservation was being made _on the OCaml stack, not the C stack_.

The fix I propose here requires an understanding of the interaction between the C and OCaml stacks. The startup process for a native program begins - in C - with `caml_main` which brings up the runtime system. It then hands off to `caml_start_program` (in `amd64.S`). Its job is to switch to the newly allocated OCaml stack and execute the program's OCaml entry point on it. The pointer to the C stack is stored in `Caml_state` and then the program can neatly flip between the two stacks as necessary (storing the pointer to resume to on the other stack in a book-keeping `struct` either in `Caml_state` or on the C stack itself). The complexity arises when a C function called from OCaml itself wants to execute OCaml code (i.e. the `caml_callback` family of functions). This book-keeping is managed by `struct c_stack_link`, which is declared in `fiber.h` but only ever created on the C stack by the functions in `amd64.S`. These structs form a linked list and, importantly, when running OCaml code, the C stack _always_ ends with one (this allows `caml_c_call` et al to be able to backup the OCaml stack pointer).

The second commit here capitalises on this separation of OCaml and C stacks. In OCaml 5, the C stack is only ever used to call C functions. The fix therefore is to include the shadow store reservation in `struct c_stack_link` itself (the ABI does not require it to be initialised or cleaned, so it's simply memory addressing arithmetic). This removes all the need for the native backend to have to care at all about the shadow store (logic for it in both `asmcomp/amd64/emit.mlp` and `asmcomp/amd64/proc.ml` is removed). This eliminates the wasted allocation on the OCaml stack. The actual calling of C functions is also simplified - after the sequence `SWITCH_OCAML_TO_C` in `amd64.S`, the C stack is now exactly ready to call a C function. A small amount of care is needed in `caml_c_call_stack_arguments` to put the arguments _above_ the shadow store, but otherwise the only time the shadow store has to be reserved is when C calls OCaml (i.e. the less frequent case - an OCaml program more frequently calls into C).

dra27 committed:
Fix register metadata for Win64 ABI

rbx is normally preserved in the Win64 ABI (as for Unix), but OCaml 5
uses it to save the OCaml stack pointer for direct (noalloc) calls. The
fact rbx is clobbered at C call was incorrectly missing in the Win64
register descriptions.

dra27 committed:
Place Win64 ABI shadow store in c_stack_link

The Win64 ABI requires the caller to reserve a fixed 32 bytes of stack
space prior to the call for the function to spill the four argument
registers into (if it chooses).

There are three ways C functions are called by OCaml programs. Under
Windows, two of them were needlessly wasting space reserving this twice
and the third was not reserving it at all.

This implementation capitalises on the fact that OCaml 5 no longer uses
the C stack for OCaml code and so adds the shadow store to the
c_stack_link struct. When in OCaml code, the C stack is therefore left
in exactly the required state to call a C function, which removes the
need for the stack pointer computations from the backend.
Simultaneously, this also fixes a bug, as the shadow store previously
was not being reserved at all when calling noalloc C functions.

kayceesrk commented:
Thanks. I've read the description and learnt a bit about the Windows x64 calling convention through that. The relevant bits from [x64 calling convention on Windows](https://learn.microsoft.com/en-us/cpp/build/x64-calling-convention?view=msvc-170) are:

> The x64 Application Binary Interface (ABI) uses a four-register fast-call calling convention by default. Space is allocated on the call stack as a shadow store for callees to save those registers.

> For prototyped functions, all arguments are converted to the expected callee types before passing. The caller is responsible for allocating space for the callee's parameters. The caller must always allocate sufficient space to store four register parameters, even if the callee doesn't take that many parameters. This convention simplifies support for unprototyped C-language functions and vararg C/C++ functions. For vararg or unprototyped functions, any floating point values must be duplicated in the corresponding general-purpose register. 

Since the x64 ABI calling convention only applies on the C side, we only need to ensure that enough space is reserved when calling from OCaml to C. Reserving the space for this in `struct c_stack_link` is clever and avoids repeated stack manipulation. 

> Any parameters beyond the first four must be stored on the stack after the shadow store before the call. Vararg function details can be found in [Varargs](https://learn.microsoft.com/en-us/cpp/build/x64-calling-convention?view=msvc-170#varargs). Unprototyped function information is detailed in [Unprototyped functions](https://learn.microsoft.com/en-us/cpp/build/x64-calling-convention?view=msvc-170#unprototyped-functions).

It is not clear to me what "after the shadow store before the call" means. Does that mean that the top of the stack contains the reserved space for the shadow store and then we have the arguments on the stack. (It must be since the tests are passing :-) ). 

Due to this requirement, specialisation in `caml_c_call_stack_args` is necessary, which first removes the space reserved for shadow store (originally reserved in `caml_start_program`), pushes the stack arguments, and then reserves the space for the shadow store. After the C call returns, the stack pointer is reset to the saved value `Caml_state(c_stack)` which includes the space for the shadow store. 

The code looks right to me. 

nit: The Windows x86 calling convention calls the reserved space as "Shadow Store" not "Shadow Stack". I was confused for a bit since "Shadow Stack" means something else. Would be useful to fix the description above.

dra27 commented:
Thanks, @kayceesrk! Rebased to fix the conflict on `Changes`, and to update the nit...

> nit: The Windows x86 calling convention calls the reserved space as "Shadow Store" not "Shadow Stack". I was confused for a bit since "Shadow Stack" means something else. Would be useful to fix the description above.

A very good point! I changed the (unused) field name in `caml/fiber.h` similarly.

> It is not clear to me what "after the shadow store before the call" means.

I puzzled a ~bit~lot over this too - it's made slightly less unclear by the stack diagram in [x64 stack usage](https://learn.microsoft.com/en-us/cpp/build/stack-usage) (although this diagram I found unclear for other reasons!!):
![x64 stack usage diagram](https://learn.microsoft.com/en-us/cpp/build/media/vcamd_conv_ex_5.png?view=msvc-170)

From this I included that for a growing down stack that "after" meant at a higher address. It has a certain amount of logic, in that it means if the register parameters are spilled to their home locations (which debug prologues at `-Od` seem to), then the parameters appear in descending order on the stack.

xavierleroy commented:
I don't have time to review this PR right now, but would  like to do so at some point.  I agree that this kind of stack space that must be reserved by the caller is 1- bad ABI design, but 2- something that can be handled more nicely in OCaml 5 than in OCaml 4 because it is now clear that the extra space need not be reserved on the OCaml stack, only on the C stack just before the C call.  I made a similar suggestion for the s390x port #11712 but haven't checked how the proposed s390x port does it.

kayceesrk reviewed (approved):
The code looks good to me.

dra27 commented:
Thanks, @kayceesrk!

@xavierleroy - no problem at all, I’ll keep it rebased.

xavierleroy commented on asmcomp/amd64/proc.ml:
```diff
@@ -297,7 +297,7 @@ let destroyed_at_c_call =
   if win64 then
     (* Win64: rbx, rsi, rdi, r12-r15, xmm6-xmm15 preserved *)
     Array.of_list(List.map phys_reg
-      [0;4;5;6;7;10;11;12;
+      [0;1;4;5;6;7;10;11;12;
```
Could you update the comments two lines above?

xavierleroy commented on runtime/amd64.S:
```diff
@@ -635,13 +630,19 @@ CFI_STARTPROC
     /* Make the alloc ptr available to the C code */
         movq    %r15, Caml_state(young_ptr)
     /* Copy arguments from OCaml to C stack */
+#if defined(SYS_mingw64) || defined (SYS_cygwin)
+        addq    $32, %rsp
+#endif
```
I understand what you're doing here, but maybe a comment is warranted, something like
"Temporarily discard the reserved 32 bytes at bottom of stack while copying the arguments".

xavierleroy reviewed (commented):
I finally found time to read this PR, thanks for your patience.

I'm 100% in favor of not reserving the 32 extra bytes on the OCaml stack and reserving them on the C stack only, since that's the only place we need them.  Plus, it makes the OCaml code in proc.ml and emit.mlp simpler, that's great!

I'm only 50% in favor of incorporating the 32 extra bytes in the c_stack_link structure.  First, it puts more machine dependencies in <caml/fiber.h>.  Second, it streamlines the code for `caml_c_call` but makes `caml_c_call_stack_args` a bit more clumsy.  Just bracketing calls to C functions with "allocate whatever extra stack space is required" and "free the extra space" feels clearer to me.

What do you (@dra27 and others) think about this?

xavierleroy commented:
I'm pretty sure the s390x port (#11712) can benefit from the same tricks, since the ABI mandates 160 reserved bytes at the bottom of the stack before calling a C function.  Note that making these 160 bytes part of `c_stack_link` means more ifdefs in `<caml/fiber.h>`, though.

kayceesrk commented:
> Just bracketing calls to C functions with "allocate whatever extra stack space is required" and "free the extra space" feels clearer to me.

I agree that this would be cleaner. This approach has a small runtime cost, which the current method avoids in almost all of the cases; `noalloc` C calls and `caml_c_call` will also need to allocate and free this space. Given that allocation and free here is just the manipulation of the stack pointer, the cost is likely to be insignificant. 

I'm in favour of this approach.

dra27 commented:
Thanks for the review, @xavierleroy! I think it's doable both without the bracketing and without altering `struct c_stack_link` (I started it last night) - I'll push that and you can see what you think, and then otherwise we do the bracketed version. Generated code really is peppered with these "brackets" coming from ocamlopt-generated noalloc calls and it was nice to make the generated code "consistent" between Win32/Linux (acknowledging of course that neither the increased code size nor the stack manipulations should have any remotely measurable performance impact!).

I'm not sure if I'll push this before going on vacation next week, but there should be something by a fortnight's time.

xavierleroy commented:
> This approach has a small runtime cost, which the current method avoids in almost all of the cases; noalloc C calls and caml_c_call will also need to allocate and free this space. 

That's right.  For "noalloc" calls, it's nice to avoid the alloc/free dance, as it increases code size and degrades code cache performance.  For caml_c_call it's less of an issue, since it's hot code that should always be in the cache.

> I'm not sure if I'll push this before going on vacation next week, but there should be something by a fortnight's time.

Take your time!  Will gladly review after your vacation.

dra27 commented:
OK, this is obviously quite important to fix prior to 5.1. I'm likely to lag responding to further things with this, but here's where things are.

The main PR is updated to fix the rbx comment in `Proc.destroyed_at_c_call`.

I've toyed with a few alternative approaches for dealing with the shadow store, here are the links to the commits and what look to me to be the pros and cons of each.

- The current PR places it in the definition of `struct c_stack_link` in `fiber.h`. The downside here is that this detail is exposed to C. On the other hand, this structure is strictly internal (like some of the details already in `stack.h`) and the changes needed in the assembly stubs seem the simplest.
- https://github.com/dra27/ocaml/commit/46a1db1686bdb145c7535c075be7f534bab8d32a fixes the "bracketing" only. It has a clear code-benefit that the assembly stubs do not require altering. It comes at a very noticeable code size cost, though - just looking at the compiler distribution, this increases code size in executables and native libraries by ~0.4% (biggest was ~1.3% for `unix.a`).
- https://github.com/dra27/ocaml/commit/dfc953116d1819710af6f9d338b0ded1c1db5ae6 keeps the idea that the C stack is left ready with the shadow store reserved, but does this entirely in `amd64.S`, with no alteration to the `c_stack_link` struct. This has the side-effect that `Caml_state->c_stack` remains the `c_stack_link` structure (which is the _only_ thing the C part of the runtime needs from it), but it's not necessarily the _actual_ C stack pointer. However, that fact is contained within the assembly stubs only. The manipulation in the assembly stubs is much more involved as a result.
- https://github.com/dra27/ocaml/commit/ba101b0cac510431f01338b6c3687a1a00e3a6dd is a further variation which keeps `Caml_state->c_stack` as being the actual C stack pointer but provides a mechanism for `fiber.c` to map from a pointer to the C stack to `struct c_stack_link` in `stack.h` (in a similar vein to existing mechanisms for gc_regs, etc.). This is slightly simpler than the previous effort, because `Caml_state->c_stack` is now reliably the C stack - the complexity for maintaining the links is solely in the three functions (`caml_start_program`, `caml_callback*` and `caml_try_realloc_stack`). It's a bit fiddly, though - there may well be a better trick for dealing with the final `NULL` pointer cases as well.

At this stage, these three new commits are passing the testsuite, but they have _not_ been as battle-tested as the main PR (which has been tested with the msvc64 port and verified with opam packages).

samwgoldman commented:
Will single-domain OCaml 5 programs run into this issue? I'm trying to understand if I should wait until OCaml 5.1 to upgrade from OCaml 4.14 for a program which runs on Windows. Thanks!

gasche commented:
Yes, this also affects single-domain programs (the problem here will be the interleaving of calls between C and OCaml, so typically calling certain external primitives from OCaml).

jmid commented:
FYI, David has shared a version of the Windows mingw opam package already patched with this PR to prevent 5.0.0 Windows users from getting bitten by this bug: https://discuss.ocaml.org/t/pre-ann-installing-windows-ocaml-5-0-0-in-opam/11150

xavierleroy committed:
AMD64 emitter: replace a trivial `lea` by a `mov`

xavierleroy reviewed (approved):
Sorry for the late review.  Everything looks good to me.  I took the liberty to push a trivial code improvement (use `mov` instead of `lea` in one place).  The `#ifdef` in `<caml/fiber.h>` seems unavoidable given that we don't want to generate stack adjustments around calls.

I'm running some more CI precheck and will merge when it's positive.

xavierleroy commented:
Merged and cherry-picked to 5.1.

kayceesrk commented:
Thanks Xavier.
